sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java

```diff
@@ -0,0 +1,311 @@
+package com.azure.json;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Iterator;
+
+/**
+ * Class representing the JSON array type
+ */
+public class JsonArray extends JsonDataStructure {
+
+    public JsonArray() {
+        super();
+    }
+
+    public JsonArray(JsonReader reader) throws IOException {
+        super();
+        this.build(reader);
+    }
+    /**
+     * Stores the JsonElements nested in this JsonArray object.
+     * Each of these elements should be one of the following valid JSON types:
+     * object, array, string, number, boolean, and null.
+     */
+
+
+    List<JsonElement> elements = new ArrayList<>();
+
+
+    /**
+     * Adds a JsonElement object as a nested element within the JsonArray object,
+     * appending it to the end.
+     *
+     * @param element a JsonElement object representing one of the valid JSON
+     * types: object, array, string, number, boolean, and null
+     * @return JsonArray object representing the new state of the JsonArray object
+     * after the addition of the new JsonElement object appended.
+     */
+    public JsonArray addElement(JsonElement element) {
+        this.elements.add(element);
+        return this;
+    }
+
+    /**
+     * Adds a JsonElement object as a nested element within the JsonArray object
+     * at a particular index, shifting the existing elements from that index to
+     * the right.
+     *
+     * @param element a JsonElement object representing one of the valid JSON
+     * types: object, array, string, number, boolean, and null
+     * @param index the element index within the JsonArray to add the new JsonElement
+     * object and where to shift from
+     * @return JsonArray object representing the new state of the JsonArray object
+     * after the addition of the new JsonElement object.
+     * TODO: throws error(s) relating to the index in respect to the current state
+     * of the elements ArrayList
+     */
+    public JsonArray addElement(int index, JsonElement element) {
```

Comment:
Something that will need to be reconciled with all the `index`-based APIs is the behavior around gaps between elements. Ex. `[JsonElement, null, null, JsonElement, null, null, JsonElement]`, what is the serialization behavior here? Do null objects become JSON nulls? Do we skip null objects in serialization?

Reply:
Right, that's one step but I was also thinking about the case where the internal List has a size of 10 (default size when `ArrayList` is instantiated) and then `addElement(2, element)` is called. This would leave a list with `[null, null, element, null, ...]`.

Reply:
I have added the following code changes to the JsonArray methods: 

```cpp
    public JsonArray addElement(JsonElement element) throws IllegalArgumentException {
        if(element == null) { 
            throw new IllegalArgumentException("Cannot add a null JsonElement to the JsonArray. Use a JsonNull object to represent a valid JSON null value."); 
        }
```

```cpp
    public JsonArray addElement(int index, JsonElement element) throws IllegalArgumentException {
        if(element == null) { 
            throw new IllegalArgumentException("Cannot add a null JsonElement to the JsonArray. Use a JsonNull object to represent a valid JSON null value."); 
        }
```

```cpp
    public JsonArray setElement(int index, JsonElement element) throws IllegalArgumentException {
        if(element == null) { 
            throw new IllegalArgumentException("Cannot set a null JsonElement in the JsonArray. Use a JsonNull object to represent a valid JSON null value."); 
        }
```

In combination with changing the initialisation of the JsonArray.elements field to size 0: 

```cpp
    private List<JsonElement> elements = new ArrayList<>(0);
```
This ensures that at all times that all elements in the JsonArray.elements List are instantiated JsonElement objects and means that the only valid way to represent null values in the JsonArray is with instantiated JsonNull objects.